lib/matplotlib/backend_bases.py

```diff
@@ -3195,3 +3207,560 @@ def zoom(self, *args):
+            - 'tool_added_event'
+
+            For every tool added a new event is created
+
+            - 'tool_trigger_TOOLNAME`
+              Where TOOLNAME is the id of the tool.
+
+        func : function
+            Function to be called with signature
+            def func(event)
+        """
+        return self._callbacks.connect(s, func)
+
+    def nav_disconnect(self, cid):
+        """
+        Disconnect callback id cid
+
+        Example usage::
+
+            cid = navigation.nav_connect('tool_trigger_zoom', on_press)
+            #...later
+            navigation.nav_disconnect(cid)
+        """
+        return self._callbacks.disconnect(cid)
+
+    def message_event(self, message, sender=None):
+        """ Emit a `NavigationMessageEvent`"""
+        if sender is None:
+            sender = self
+
+        s = 'tool_message_event'
+        event = NavigationMessageEvent(s, sender, message)
+        self._callbacks.process(s, event)
+
+    @property
+    def active_toggle(self):
+        """
+        Toggled Tool
+
+        **dict** :  Currently toggled tools
+        """
+
+        return self._toggled
+
+    def get_tool_keymap(self, name):
+        """
+        Get the keymap associated with the specified tool
+
+        Parameters
+        ----------
+        name : string
+            Name of the Tool
+
+        Returns
+        ----------
+        list : list of keys associated with the Tool
+        """
+
+        keys = [k for k, i in six.iteritems(self._keys) if i == name]
+        return keys
+
+    def _remove_keys(self, name):
+        for k in self.get_tool_keymap(name):
+            del self._keys[k]
+
+    def set_tool_keymap(self, name, *keys):
+        """
+        Set the keymap to associate with the specified tool
+
+        Parameters
+        ----------
+        name : string
+            Name of the Tool
+        keys : keys to associate with the Tool
+        """
+
+        if name not in self._tools:
+            raise KeyError('%s not in Tools' % name)
+
+        self._remove_keys(name)
+
+        for key in keys:
+            for k in validate_stringlist(key):
+                if k in self._keys:
+                    warnings.warn('Key %s changed from %s to %s' %
+                                  (k, self._keys[k], name))
+                self._keys[k] = name
+
+    def remove_tool(self, name):
+        """
+        Remove tool from `Navigation`
+
+        Parameters
+        ----------
+        name : string
+            Name of the Tool
+        """
+
+        tool = self.get_tool(name)
+        tool.destroy()
+
+        # If is a toggle tool and toggled, untoggle
+        if getattr(tool, 'toggled', False):
+            self.tool_trigger_event(tool, 'navigation')
+
+        self._remove_keys(name)
+
+        s = 'tool_removed_event'
+        event = ToolEvent(s, self, tool)
+        self._callbacks.process(s, event)
+
+        del self._tools[name]
+
+    def add_tools(self, tools):
```

Comment:
Iirc, the biggest gripe I had lay in doing this in ``NavigationBase.__init__``, because of some deadlock in converting to an event-based paradigm, you needed the ``NavigationBase`` class created before you could create your tools (or something like that), and thus shouldn't create the tools from init.  A long time ago in the PR before this one (or even the one before that) so my memory a lil' hazy on details.  The code has changed a lot since then, so not too sure if my original bad feeling against it still stand.

I would still vote to keep it this way though, in terms of use case for having this method here, I see it as a nice helper method.  Writing out the loop every time you want to use it seems like a bit of a waste of manpower to me, when we can encapsulate it here.  Speaking personally, if I were to add/leave out a letter by mistake, I would curse myself; if I were to type out the extra loop I would curse mpl for not including a helper method, perhaps I have gotten too used to frameworks taking away such common loops.